pkg/utils/etcdutil/etcdutil.go

```diff
@@ -350,3 +353,240 @@ func InitOrGetClusterID(c *clientv3.Client, key string) (uint64, error) {
+)
+
+// LoopWatcher loads data from etcd and sets a watcher for it.
+type LoopWatcher struct {
+	ctx            context.Context
+	wg             *sync.WaitGroup
+	client         *clientv3.Client
+	key            string
+	name           string
+	forceLoadCh    chan struct{}
+	isLoadedCh     chan error
+	putFn          func(*mvccpb.KeyValue) error
+	deleteFn       func(*mvccpb.KeyValue) error
+	postEventFn    func() error
+	opts           []clientv3.OpOption
+	loadTimeout    time.Duration
+	loadRetryTimes int
+	loadBatchSize  int64
+}
+
+// NewLoopWatcher creates a new LoopWatcher.
+func NewLoopWatcher(ctx context.Context, wg *sync.WaitGroup, client *clientv3.Client, name, key string,
+	putFn, deleteFn func(*mvccpb.KeyValue) error, postEventFn func() error, opts ...clientv3.OpOption) *LoopWatcher {
+	return &LoopWatcher{
+		ctx:            ctx,
+		client:         client,
+		name:           name,
+		key:            key,
+		wg:             wg,
+		forceLoadCh:    make(chan struct{}, 1),
+		isLoadedCh:     make(chan error, 1),
+		putFn:          putFn,
+		deleteFn:       deleteFn,
+		postEventFn:    postEventFn,
+		opts:           opts,
+		loadTimeout:    defaultLoadDataFromEtcdTimeout,
+		loadRetryTimes: defaultLoadFromEtcdRetryTimes,
+		loadBatchSize:  defaultLoadBatchSize,
+	}
+}
+
+// StartWatchLoop starts a loop to watch the key.
+func (lw *LoopWatcher) StartWatchLoop() {
+	defer logutil.LogPanic()
+	defer lw.wg.Done()
+
+	ctx, cancel := context.WithTimeout(lw.ctx, lw.loadTimeout)
+	defer cancel()
+	watchStartRevision := lw.initFromEtcd(ctx)
+
+	log.Info("start to watch loop", zap.String("name", lw.name), zap.String("key", lw.key))
+	for {
+		select {
+		case <-lw.ctx.Done():
+			close(lw.forceLoadCh)
+			log.Info("server is closed, exit watch loop", zap.String("name", lw.name), zap.String("key", lw.key))
+			return
+		default:
+		}
+		nextRevision, err := lw.watch(lw.ctx, watchStartRevision)
+		if err != nil {
+			log.Error("watcher canceled unexpectedly and a new watcher will start after a while for watch loop",
+				zap.String("name", lw.name),
+				zap.String("key", lw.key),
+				zap.Int64("next-revision", nextRevision),
+				zap.Time("retry-at", time.Now().Add(watchEtcdChangeRetryInterval)),
+				zap.Error(err))
+			watchStartRevision = nextRevision
+			time.Sleep(watchEtcdChangeRetryInterval)
+		}
+	}
+}
+
+func (lw *LoopWatcher) initFromEtcd(ctx context.Context) int64 {
+	var (
+		watchStartRevision int64
+		err                error
+	)
+	ticker := time.NewTicker(defaultLoadFromEtcdRetryInterval)
+	defer ticker.Stop()
+
+	for i := 0; i < lw.loadRetryTimes; i++ {
+		watchStartRevision, err = lw.load()
+		failpoint.Inject("loadTemporaryFail", func(val failpoint.Value) {
+			if maxFailTimes, ok := val.(int); ok && i < maxFailTimes {
+				err = errors.New("fail to read from etcd")
+				failpoint.Continue()
+			}
+		})
+		if err == nil {
+			if deadline, ok := ctx.Deadline(); ok && deadline.After(time.Now()) {
```

Comment:
What is it used for?

Reply:
when test `delayLoad` to check whether timeout

Reply:
Then it's just for test?